File: java/src/main/java/algorithm/utils/errors/ShapeError.java
package algorithm.utils.errors;

import java.util.List;

/**
 * Abstract class for all shape-related errors.
 */
public class ShapeError extends Exception {
    protected final String shape;
    protected final List<Double> dimensions;
    protected final String reason;

    public ShapeError(String shape, List<Double> dimensions, String reason) {
        this.shape = shape;
        this.dimensions = dimensions;
        this.reason = reason;
    }

    @Override
    public String toString() {
        String name = this.getClass().getSimpleName();
        String message = getMessage();
        return String.format("%s: %s", name, message);
    }

    @Override
    public String getMessage() {
        String shape = this.shape.toUpperCase();
        String dimLabel = dimensions.size() == 1 ? "dimension" : "dimensions";
        String dimensions = formatDimensions();
        String reason = this.reason;
        return String.format("%s with %s %s is invalid. %s", shape, dimLabel, dimensions, reason);
    }

    private String formatDimensions() {
        if (dimensions == null || dimensions.isEmpty()) {
            return "no dimensions";
        }
        switch (dimensions.size()) {
            case 1:
                return String.valueOf(dimensions.get(0));
            case 2:
                return String.format("%s and %s", dimensions.get(0), dimensions.get(1));
            default:
                Double last = dimensions.get(dimensions.size() - 1);
                String restStr = dimensions.subList(0, dimensions.size() - 1)
                    .stream()
                    .map(String::valueOf)
                    .reduce((a, b) -> a + ", " + b)
                    .orElse("");
                return String.format("%s, and %s", restStr, last);
        }
    }
}

File: python/src/algorithm/utils/errors/shape_error.py
from abc import ABC, abstractmethod


class ShapeError(ABC, Exception):
    """
    Base exception class for all shape-related errors.
    """

    @abstractmethod
    def __init__(self, shape: str, dimensions: list[float], reason: str):
        self.shape = shape
        self.dimensions = dimensions
        self.reason = reason
        super().__init__(self.__str__)

    def _format_dimensions(self):
        match self.dimensions:
            case [d]:
                return str(d)
            case [d1, d2]:
                return f"{d1} and {d2}"
            case [*rest, last]:
                return f"{', '.join(map(str, rest))}, and {last}"
            case _:
                return "no dimensions"

    def _message(self):
        shape = self.shape.upper()
        dim_label = "dimension" if len(self.dimensions) == 1 else "dimensions"
        dimensions = self._format_dimensions()
        reason = self.reason
        return f"{shape} with {dim_label} {dimensions} is invalid. {reason}"

    def __str__(self):
        name = self.__class__.__name__
        message = self._message()
        return f"{name}: {message}"
